### PR TITLE
Centralize install code

### PR DESCRIPTION
**Pull request by lavalamp**
Trying to figure out a way to do this that makes the changes as painless to roll out as possible. This is going to be a multi-step process...

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30565)

### CODE REVIEW

**Comment by lavalamp** [comment]
OK, I think this is ready for review. I've structured this so that I can refactor in smallish chunks.

**Comment by lavalamp** [comment]
First commit is #30663.

**Comment by smarterclayton** [comment]
I knew it was only a matter of time before we got factories.

**Comment by nikhiljindal** [comment]
I assume this is not for 1.4, but want to confirm

**Comment by lavalamp** [comment]
Factory produced groups are better than artisanal hand-crafted groups!

@nikhiljindal it would be great to get in. It should be pretty easy to review at this point? Hopefully the tests pass now.

**Comment by lavalamp** [comment]
OK, I think I fixed the boilerplate & test for real this time.

File: pkg/api/testapi/testapi.go [hunk 1]
```diff
@@ -15,6 +15,12 @@ limitations under the License.
 */
 
 // Package testapi provides a helper for retrieving the KUBE_TEST_API environment variable.
+//
+// TODO(lavalamp): this package is a huge disaster at the moment. I intend to
+// refactor. All code currently using this package should change:
+// 1. Declare your own registered.APIGroupRegistrationManager in your own test code.
+// 2. Import the relevant instal packages.
```
Comment (nikhiljindal) [inline pkg/api/testapi/testapi.go]: install

File: pkg/apimachinery/announced/announced.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package announced contains tools for announcing API group factories. This is
+// distinct from registration (in the 'registered' package) in that it's safe
+// to announce every possible group linked in, but only groups requested at
+// runtime should be registered. This package contains both a registry, and
+// factory code (which was formerly copy-pasta in every install package).
```
Comment (nikhiljindal) [inline pkg/apimachinery/announced/announced.go]: pasta -> paste

File: pkg/apimachinery/announced/announced.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package announced contains tools for announcing API group factories. This is
+// distinct from registration (in the 'registered' package) in that it's safe
+// to announce every possible group linked in, but only groups requested at
```
Comment (nikhiljindal) [inline pkg/apimachinery/announced/announced.go]: Isnt that the difference between registered and enabled? You can register all groups, but group will be added to Scheme only when it is enabled.

File: pkg/apimachinery/announced/announced.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+// users of the registry can select which groups and versions they'd actually
+// like to register with an APIRegistrationManager.
+//
+// The key is the group name. After initialization, this should be treated as
+// read-only. It is implemented as a map from group name to group factory, and
+// it is safe to use this knowledge to manually pick out groups to register
+// (e.g., for testing).
+type APIGroupAnnouncementRegistry map[string]*GroupMetaFactory
+
+func (gfr APIGroupAnnouncementRegistry) group(groupName string) *GroupMetaFactory {
```
Comment (nikhiljindal) [inline pkg/apimachinery/announced/announced.go]: Why gfr? Should it be gar?

File: pkg/apimachinery/announced/group_factory.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/apimachinery"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
+)
+
+type ApplyToScheme func(*runtime.Scheme) error
+type VersionToSchemeFunc map[string]ApplyToScheme
```
Comment (nikhiljindal) [inline pkg/apimachinery/announced/group_factory.go]: its not a func?

File: pkg/apimachinery/announced/group_factory.go [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+// GroupMetaFactoryArgs contains the group-level args of a GroupMetaFactory.
+type GroupMetaFactoryArgs struct {
+	GroupName              string
+	VersionPreferenceOrder []string
+	ImportPrefix           string
+
+	RootScopedKinds sets.String // nil is allowed
+	IgnoredKinds    sets.String // nil is allowed
+
+	// May be nil if none
```
Comment (nikhiljindal) [inline pkg/apimachinery/announced/group_factory.go]: incomplete sentence?

File: pkg/apimachinery/announced/group_factory.go [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+		prioritizedVersions = append(
+			prioritizedVersions,
+			unversioned.GroupVersion{
+				gmf.GroupArgs.GroupName,
+				v,
+			},
+		)
+	}
+
+	availableVersions := []unversioned.GroupVersion{}
```
Comment (nikhiljindal) [inline pkg/apimachinery/announced/group_factory.go]: Is unprioritizedVersions a better name.
Whatever the name, add a comment saying this is the array of versions that are registered, but are not in the prioritized versions list.

File: pkg/apimachinery/announced/group_factory.go [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+			return fmt.Errorf("found %v/%v in group %v?", v.GroupName, v.VersionName, gmf.GroupArgs.GroupName)
+		}
+		if pvSet.Has(v.VersionName) {
+			pvSet.Delete(v.VersionName)
+			continue
+		}
+		availableVersions = append(availableVersions, unversioned.GroupVersion{v.GroupName, v.VersionName})
+	}
+	if len(availableVersions) != 0 {
+		glog.Warningf("group %v has unprioritized versions %#v. They will have an arbitrary preference order!", gmf.GroupArgs.GroupName, availableVersions)
```
Comment (nikhiljindal) [inline pkg/apimachinery/announced/group_factory.go]: Error?

**Comment by nikhiljindal** [comment]
There is a lot of complexity here: announced, registered and enabled group versions; factory for announcing, registering and enabling group versions; a registry of such factories; APIRegistrationManager.
Its not easy to wrap your head around all of this.

Do we need all of these layers? Any layer that we dont really need right now?

**Comment by nikhiljindal** [comment]
cc @kubernetes/sig-api-machinery

**Comment by lavalamp** [comment]
Yeah, this is complex but the complexity is already there. This makes it
centralized complexity instead of distributed complexity. It can still be
significantly improved but I want to make the improvements in 2 packages
and not 20.

File: pkg/apimachinery/announced/announced.go (hunk 2 above)
Comment (lavalamp) [inline pkg/apimachinery/announced/announced.go]: This is actually deliberate-- copy-pasta is for code that's been copied and pasted and is now spaghetti ;)

if you google "copypasta code" you'll see that I'm not the first to use it this way.

File: pkg/apimachinery/announced/group_factory.go (hunk 5 above)
Comment (lavalamp) [inline pkg/apimachinery/announced/group_factory.go]: Parse as "Version To SchemeFunc"-- I'll change the name of ApplyToScheme to make this clearer

File: pkg/apimachinery/announced/group_factory.go (hunk 8 above)
Comment (lavalamp) [inline pkg/apimachinery/announced/group_factory.go]: Hm, maybe? I guess I'd rather wait and see what people end up doing with this.

Comment (lavalamp) [inline pkg/apimachinery/announced/group_factory.go]: Actually it's only problematic if there's more than one of these, now that I think about it :)

**Comment by nikhiljindal** [comment]
> Yeah, this is complex but the complexity is already there. This makes it
> centralized complexity instead of distributed complexity. 

Yes the complexity was still there. But now there are new concepts: announced group versions, registery, factory.

> It can still be
> significantly improved but I want to make the improvements in 2 packages
> and not 20.

Agreed, thats reasonable.


A few points:
* GroupMetaFactory does not sound like the right name: With that name, I expect it to produce a GroupMeta, but what it does is register and enable the right group versions in APIRegistrationManager. Should it be APIRegistrationManager{Generator, Builder, Factory}?
* Announce: All it does is add a group version to the registry. How about calling it Add instead of announce? 

Without having worked out exact details, a high level flow that makes sense is: you *Add* your group versions to a APIRegistrationManagerFactory, the factory produces APIRegistrationManager which knows which group versions and resources are enabled and then all other code just uses APIRegistrationManager.
Right now, we have code in install.go on how to update APIRegistrationManager. With this change, they will just add their group version to the factory, which has the common code to update APIRegistrationManager.

**Comment by lavalamp** [comment]
here's two commits with the first batch of fixes. Still pondering your name suggestions.

File: pkg/apimachinery/announced/announced.go (hunk 3 above)
Comment (lavalamp) [inline pkg/apimachinery/announced/announced.go]: Added a section here and in the registered package.

**Comment by lavalamp** [comment]
re: naming

Today, the API group registration manager has two concepts:
* Registration, which I want to start calling "announcing" and move to the announcement manager in the new package.
* Enabling, which I want to leave in the registered package, and start calling "registration".

If you look at GroupMetaFactory, it does actually produce GroupMeta objects, but it's tightly coupled with the registration manager. The idea I'm trying to present here is that you "announce" your group's components, those get combined into a GroupMetaFactory, and then anyone can use the factory to place the correct GroupMeta in their APIGroupRegistrationManager.

**Comment by lavalamp** [comment]
OK, I changed the name to "APIGroupFactoryRegistry" and added comments in several places justifying the "GroupMetaFactory" name. Added other misc comments. Squashed.

Thanks for the comments, can you live with it this way until the next installment? :)

**Comment by nikhiljindal** [comment]
LGTM, thanks

**Comment by pwittrock** [comment]
Added do not merge label to sync up with Daniel on this before it goes in

**Comment by luxas** [comment]
Please say this is going to make v1.4...

**Comment by pwittrock** [comment]
Marking this as P0 because rebasing it has created a lot of work for Daniel and he has been trying to get it in since pre-1.3, but pulled off of it to perform heroics on the release.  Moving it to milestone v1.5.

**Comment by pwittrock** [comment]
@luxas I talked with Daniel and I would like this to be the first PR for v1.5.  I don't believe putting it in v1.4 serves the interest of the release.  I know realize this is super painful and thank you for your patience.  Msg me on slack if you want to chat further about this decision.

**Comment by lavalamp** [comment]
(Toggled the labels--forgot to add that milestone to the blacklist. Fixed now.)

**Comment by lavalamp** [comment]
@k8s-bot test this issue #IGNORE

**Pull request merged**